cocos/audio/openharmony/UrlAudioPlayer.cpp

```diff
@@ -0,0 +1,358 @@
+            auto it = __playerContainer.find(player);
+            if (it != __playerContainer.end())
+            {
+                UrlAudioPlayer *audioPlayer = (UrlAudioPlayer *)it->second;
+                audioPlayer->playEventCallback();
+            }
+        }
+    }
+}
+
+void UrlAudioPlayer::playEventCallback(){
+    std::shared_ptr<bool> isDestroyed = _isDestroyed;
+    auto func = [this, isDestroyed]() {
+        // If it was destroyed, just return.
+        if (*isDestroyed)
+        {
+            ALOGV("The UrlAudioPlayer (%p) was destroyed!", this);
+            return;
+        }
+
+        //Note that It's in the caller's thread (Cocos Thread)
+        // If state is already stopped, ignore the play over event.
+        if (_state == State::STOPPED)
+        {
+            return;
+        }
+
+        setState(State::OVER);
+        if (_playEventCallback != nullptr)
+        {
+            _playEventCallback(State::OVER);
+        }
+
+        ALOGV("UrlAudioPlayer (%p) played over, destroy self ...", this);
+        destroy();
+        delete this;
+    };
+
+    if (_callerThreadId == std::this_thread::get_id())
+    {
+        func();
+    }
+    else
+    {
+        cocos2d::Application::getInstance()->getScheduler()->performFunctionInCocosThread(func);
+    }
+    
+}
+
+bool UrlAudioPlayer::prepare(const std::string &url, std::shared_ptr<AssetFd> assetFd, int32_t start, int32_t length)
+{
+    _url = url;
+    _assetFd = assetFd;
+    _playObj = OH_AVPlayer_Create();
+
+    __playerContainerMutex.lock();
+    __playerContainer[_playObj] = this;
+    __playerContainerMutex.unlock();
+
+    AVPlayerCallback callback;
+    callback.onInfo = this->onInfo;
+    OH_AVPlayer_SetPlayerCallback(_playObj, callback);
+    OH_AVPlayer_SetFDSource(_playObj, _assetFd->getFd(), start, length);
+    OH_AVErrCode code = OH_AVPlayer_Prepare(_playObj);
+    if (code == AV_ERR_OK) {
+        setState(State::INITIALIZED);
+        return true;
+    } 
+
+    ALOGE("Oops, prepare error: %d", (int)code);
+    return false;
+}
+
+void UrlAudioPlayer::rewind()
+{
+// Not supported currently. since cocos audio engine will new -> prepare -> play again.
+}
+
+void UrlAudioPlayer::setLoop(bool isLoop)
+{
+    _isLoop = isLoop;
+    int status = OH_AVPlayer_SetLooping(_playObj, _isLoop);
+    if(status != 0) {
+        ALOGE( "UrlAudioPlayer::setLoop %d failed", _isLoop ? 1 : 0);
+    }
+}
+
+bool UrlAudioPlayer::isLoop() const
+{
+    return _isLoop;
+}
+
+void UrlAudioPlayer::stopAll()
+{
+    __playerContainerMutex.lock();
+    auto temp = __playerContainer;
+    __playerContainerMutex.unlock();
+
+    std::map<OH_AVPlayer*, cocos2d::UrlAudioPlayer*>::iterator it = temp.begin();  
+    while(it != temp.end())  
+    {  
+        UrlAudioPlayer* thiz = (UrlAudioPlayer*)it->second;
+        thiz->stop();
+        it++;
+    }
+}
+
+void UrlAudioPlayer::destroy()
+{
+    if (!*_isDestroyed)
+    {
+        *_isDestroyed = true;
```

Comment:
Though _isDestroyed is a shared ptr, but the usage here is not thread safe.

Reply:
I think android should have the same issue. Do we need to add lock here?

Reply:
It depends on the usage:
- will this function be invoked in different threads?
- what will happen if `OH_AVPlayer_Reset(_playObj);` and `OH_AVPlayer_Release(_playObj);` be invoked more than one time?

Reply:
1 It is possible to be called by different threads
2 When these interfaces are called repeatedly,  error code 3 will be returned, but no exception will be caused.

Reply:
![image](https://github.com/cocos/engine-native/assets/41882593/8a94711b-75b7-48fe-acb7-2a63f118ad69)